### ISSUE TITLE
Coveralls doesn't block PR if test coverage decreased

### ISSUE THREAD

**Issue by st4lk** [lemurheavy/coveralls-public#1623]
When I'm creating new PR with decreased test coverage - coveralls doesn't block the PR although test coverage decreased (comparing to source branch). It shows that my branch in PR has "New Coverage".
If I'll push new commit to the branch associated with PR - only after that PR will be blocked by coveralls because test coverage decreased.
I'm wondering: why coveralls doesn't block the PR if coverage decreased on first commit?

![Screen Shot 2022-03-18 at 14 09 02](https://user-images.githubusercontent.com/1771042/158993351-e73e637e-4ba7-4498-beb9-ec68db110974.png)

<img width="1092" alt="decreased-test-coverage" src="https://user-images.githubusercontent.com/1771042/158993679-b193ea2a-8e52-4192-ad50-c23cf541cc92.png">

### Stack of tools I'm using
- [CircleCI](https://circleci.com/) to make the build
- [coveralls-python](https://github.com/TheKevJames/coveralls-python) to push coverage report to coveralls from CircleCI
- I've enabled all the threshold alerts in coveralls settings:
<img width="644" alt="Coveralls - Test Coverage History   Statistics 2022-03-18 14-06-46" src="https://user-images.githubusercontent.com/1771042/158993691-d99d5f82-4e9e-4750-8d5e-207778756b56.png">
- This is open source repository in github

Maybe I need to adjust some settings in coveralls.io somewhere?

**Comment by st4lk** [comment]
Looks like can be the same issue as described here:
https://github.com/lemurheavy/coveralls-public/issues/1167

**Comment by afinetooth** [comment]
Hi, @st4lk. So I hear what you're saying as far as the description of your issue, however some details are throwing me off.

First, from your screenshot above, you can see under BUILD TYPE (below the read header reading "FIRST BUILD ON...") that your build type is **push**, and not **pull_request**. Therefore, any settings related to pull requests would not apply for that build. In other words, Coveralls would not send a failing status update to Github, blocking the PR from being merged (because it's not a PR).

I believe I have identified the build you're using as your example, and it is the only build in your project with a creation date of `18 Mar 2022 10:52AM UTC`, so I believe we're looking at the same one. (It's about 2/3 way down the page [here in your build history](https://coveralls.io/repos/366886/builds?page=2).)

When I look further up the history toward more recent builds and get to the first build that _is_ a PR---a build starting with `2C8E5BAF-` created at `18 Mar 2022 12:43PM UTC`---I see that---*as a __pull_request__*---that build also had a decrease in coverage:
![Screen Shot 2022-03-18 at 11 48 05 AM](https://user-images.githubusercontent.com/918667/159065507-3174390e-f121-450f-94f2-c8e1472fa90f.png)

And that a failing status update _was_ sent for the PR.

From our database:

> id=16719988705
> description="Coverage decreased (-3.03%) to 63.636%" 
> state="failure"

And from the PR:
![Screen Shot 2022-03-18 at 11 50 03 AM](https://user-images.githubusercontent.com/918667/159065865-cd1faaec-d692-47cb-87be-42c12c595c54.png)

**Comment by st4lk** [comment]
Hey @afinetooth ,
thank you for the quick response!

I think this is what is happening:
- new branch is pushed to github
- circleci start the build
- PR is not opened yet
- circleci have run all the tests and have pushed coverage report to coveralls
- since PR is not opened - coveralls don't block it
- finally, PR is opened, but it will not be blocked by coveralls even if test coverage felt down

It looks like a common workflow, so I'm wondering what would be the best thing to do.
Obviously, there is some delay between pushing new branch to github and opening the PR for it.

Should we trigger coveralls once again when PR is actually opened so it can possibly block it?

(another approach could be to start the build by circleci only when PR is opened and not when branch is pushed, but it has some other disadvantages)

**Comment by afinetooth** [comment]
Hi, @st4lk. 

> - finally, PR is opened, but it will not be blocked by coveralls even if test coverage felt down

That's actually not true. Coveralls will send failing (blocking) status updates for all PRs when coverage falls below the thresholds in your settings.

It actually did that when the PR for your branch _was_ opened.

Here's the PR build:
https://coveralls.io/builds/47490484

And here's the blocking status update on [the PR at Github](https://github.com/ergeon/python-ci-cd/pull/14):

![Screen Shot 2022-03-21 at 10 52 59 AM](https://user-images.githubusercontent.com/918667/159334227-2d23b581-2843-4311-9494-4fd0d7e325dc.png)

_Question: Did you configure Github to block PRs with failing status updates?_

> It looks like a common workflow, so I'm wondering what would be the best thing to do.
> Obviously, there is some delay between pushing new branch to github and opening the PR for it.
> 
> Should we trigger coveralls once again when PR is actually opened so it can possibly block it?

Yes, you should configure CircleCI to build **pull_requests** and, if you're using a Coveralls integration, Coveralls will do the rest, in terms of sending failing (blocking) status updates to Github.

Unless I misunderstand your workflow, Coveralls is already designed to support your workflow. Let me explain how:

- Coveralls will generate coverage reports for **push** builds and _send status updates_ to Github for them. For example, here's the (failing) status update from Coveralls for the [**push** build described above](https://github.com/ergeon/python-ci-cd/commit/74c50815e5175691394994df5f3d9c2df1ecd38d):

![Screen Shot 2022-03-21 at 10 25 51 AM](https://user-images.githubusercontent.com/918667/159329770-abbdc313-6ae5-4737-9a2b-832b5f9904e2.png)

- When a **pull_request** is opened, Coveralls processes the PR build and _also_ sends status updates, but, __because the build is a PR at Github__, the status update has the ability to block the PR from being merged (if it's failing).

I guess I'm confused about what the proper Coveralls feature would be to handle your workflow, since Coveralls can't know a new branch is _intended_ to become a **pull_request**.

I think I understand why a failing status update would be useful for a new _feature branch_---_so developer(s) know they won't have a mergeable PR until meet coverage requirements_---however, as I say above, that feature is already there as far as Coveralls is concerned, because it's sending the failing status update.

Hope that makes sense. Let me know if I'm missing something.

**Comment by DysphoricUnicorn** [comment]
A similar thing happened for us just now: This [build](https://coveralls.io/builds/54181680) did not block the PR, even though it shows decreased coverage and is of type pull request.

Is there some setting I have to change to set the threshold to 100%?

**Comment by smridge** [comment]
I'm seeing this as well (for a private repo), which really defeats the purpose of requiring test coverage. I _assumed_ the default was comparing to the default branch and not the first build. I believe my options are to email support and see if there is something specific I'm not following.

**Comment by afinetooth** [comment]
Hi @smridge. Can you please share the Coveralls URL for your repo and 1-2 builds that are displaying the issue?

If you feel better not sharing here, you can email us at support@coveralls.io and just reference this issue.

**Comment by afinetooth** [comment]
@st4lk @DysphoricUnicorn @smridge wanted to follow up on this issue, with a workaround and planned fix, in the hope of providing some value before closing it.

From a recent, similar support case, I discovered a scenario I had not considered before in which a PR with decreased test coverage exceeding the decrease threshold setting could _fail_ to block the PR with a failing status check:

Essentially, this can happen when CI is configured to build _both_ **push** _and_ **pull_request** commits on any/all new branches. 

### Issue (Flavor 1)
Two status updates are sent whenever a **push** commit is made to an _open_ PR, one for the **push** commit, and one for the updated **pull_request** build. In this scenario, the status update for the **push** commit, which compares its coverage to a _previous commit on the same branch_, may succeed, while the status update for the **pull_request** build, which compares its coverage to the _base commit on the target branch_, may fail. Which one displays is essentially determined by a race condition (whichever completes last).

### Workaround (Flavor 1)
We recommend modifying your CI config to stop building **push** commits on all / new / potential PR branches, so that only the updated **pull_request** status updates are sent.

### Fix (Flavor 1)
We are working on a solution, which is to send up to two (possible) status updates for all PR builds: (1) _status update 1 (**pull_request**)_; and (2) _status update 2 (**push**)_, if present. Sending both updates will ensure that, if the decrease threshold is surpassed for the (pull_request) check, it will fail / block the PR.

### Issue (Flavor 2)
I am less certain about this scenario, but it reflects the first situation here described by @st4lk, which is that a **push** commit is built on a new branch intended to become a PR and, when the PR is opened, the status update that appears still relates to the **push** build. This shouldn't happen under normal circumstances, but I can imagine how it _could_: (a) the second update fails for some reason; (b) the **push** build is delayed in queue, or fails and is re-run, such that its status update is sent _after_ the **pull_request** build's; etc.

### Workaround (Flavor 2)
The same workaround should fix this, at least temporarily, as only **pull_request** builds will happen on new/PR branches.

### Fix (Flavor 2)
The same fix for Flavor 1 should address this, if you're building both **push** and **pull_request** commits on new/PR branches, even if it won't solve whatever issue (hopefully one-off) led to the **push** status update coming in last. No matter which comes when, at least _one_ status check will _fail_ (_status update (pull_request)_).

I'm going to close this. Please reopen if any aspect of your issue/concern is still open.
